### ISSUE TITLE
FR: error out when creating branches with @ in the name

### ISSUE THREAD

**Issue by jalil-salame** [jj-vcs/jj#4358]
**Is your feature request related to a problem? Please describe.**
I have accidentally created branches when I wanted to move a remote branch (e.g. `jj branch set branch@origin -r @-`). This usually happens when I want to set it to a specific change with the `-r` option, as `jj branch move` uses the `--to` flag instead of the `-r` flag.

**Describe the solution you'd like**
Fail to create branches with `@` in the name by default (I assume `git` allows creating branches with `@`s in the name so you want to still be able to use/create such branches, but not by default).

**Describe alternatives you've considered**
- Add `-r/--revision` as an alias to `--to` in `jj branch move`
- Disable branch creation on set without a flag (e.g. `jj branch set --create-if-missing new-branch`)

**Comment by bnjmnt4n** [comment]
There's some prior discussion in https://github.com/martinvonz/jj/pull/3932, but I believe the consensus is that `jj branch set` can set the branch (creating one if it doesn't already exist, and giving a warning to use `jj branch move` if the intent is not to create new branches). I guess it might be possible to add a `-r` flag to `jj branch move`, but that doesn't fully align with the `--from` and `--to` argument names as well (which is why I think `--to` was chosen).

Even so, I believe there's an error in your initial issue: you can't "move" a remote branch locally. The only way to do that is to move the local tracking branch to the desired commit (via `jj branch set/move`) and `jj git push` to push the branch and update the remote branch. (https://github.com/martinvonz/jj/issues/3588 is a feature request for pushing a specific commit directly to the given branch, which might be what you're looking for.)

**Comment by yuja** [comment]
> as `jj branch move` uses the `--to` flag instead of the `-r` flag.

Do you want `-r` specifically? or is it okay if `--to` had a short flag? We have `--from`/`--to` combinations in various commands, and it'll probably make sense to add short names for them.

**Comment by jalil-salame** [comment]
> > as `jj branch move` uses the `--to` flag instead of the `-r` flag.
> 
> Do you want `-r` specifically? or is it okay if `--to` had a short flag? We have `--from`/`--to` combinations in various commands, and it'll probably make sense to add short names for them.

Short flags would be nice

**Comment by jalil-salame** [comment]
> There's some prior discussion in #3932, but I believe the consensus is that `jj branch set` can set the branch (creating one if it doesn't already exist, and giving a warning to use `jj branch move` if the intent is not to create new branches). I guess it might be possible to add a `-r` flag to `jj branch move`, but that doesn't fully align with the `--from` and `--to` argument names as well (which is why I think `--to` was chosen).

I still believe this should be an error by default (creating a branch with an `@` in the name).

> Even so, I believe there's an error in your initial issue: you can't "move" a remote branch locally. The only way to do that is to move the local tracking branch to the desired commit (via `jj branch set/move`) and `jj git push` to push the branch and update the remote branch. (#3588 is a feature request for pushing a specific commit directly to the given branch, which might be what you're looking for.)

And it should probably be an error if I try to move a remote tracking branch c:

**Comment by yuja** [comment]
> I still believe this should be an error by default (creating a branch with an `@` in the name).

I don't think it should be an error (because `@` is a valid branch name), but I understand it's confusing. Perhaps, a warning can be emitted if a branch name contained a certain meta characters (such as `:`, `@`, `*`, `?`.)

**Comment by tim-janik** [comment]
> > I still believe this should be an error by default (creating a branch with an `@` in the name).
> 
> I don't think it should be an error (because `@` is a valid branch name), but I understand it's confusing. Perhaps, a warning can be emitted if a branch name contained a certain meta characters (such as `:`, `@`, `*`, `?`.)

We already issue warnings for invalid branch names.
But the current behavior makes it much easier to create "invalid" bookmarks (i.e. ones that cannot be exported to git) than to get rid of them. E.g.
```sh
Foo="$(ls)" # accidental contents
jj b c "$FOO" # works with warning
jj b d "$FOO" # will error, hint suggests to us exact:
```

Problems occur if something like `jj log -r @` is part of the prompt, *and* a bookmark is created with many newlines (due to an accident like above).
It would be nice if "invalid" bookmark creation was as hard as deletion, i.e. I'd like to see `jj b c "$FOO"` apply the same rule as `jj b d "$FOO"` and require an exact: prefix.

Related, bookmarks with invalid characters cannot be exported to git (which produces a warning), further info on valid git refs can be found here: https://git-scm.com/docs/git-check-ref-format
Rule 9 says "They cannot be the single character @"

**Comment by sersorrel** [comment]
There is an annoying footgun with `bookmark set` for example: if you forget that you have to use `-r` (or for other commands `-t` or `--to`...), you end up creating a bookmark called `@`, which is _extremely_ nonobvious from the output produced:

```console
$ jj b s main @
Created 1 bookmarks pointing to wsxrqwtw 7c14a42d @ main* | hm: Improve/fix jj configuration
Hint: Use -r to specify the target revision.
Hint: Consider using `jj bookmark move` if your intention was to move existing bookmarks.
```

I read "Created 1 bookmarks" as "great, it's (re)created `main` pointing at the change I want", instead of the actual "I've created `@` pointing at the change and `main` was incidentally left where it already was".

**Comment by yuja** [comment]
This should be addressed by b1eca3710c70c19adc8396a78db000913c9bb21a. The error message needs some improvement, though.
